src/core/knn.js

```diff
@@ -0,0 +1,131 @@
+'use strict';
+
+const { recursiveToJSON, checkType } = require('./util');
+const Query = require('./query');
+
+/**
+ * Class representing a k-Nearest Neighbors (k-NN) query.
+ * This class extends the Query class to support the specifics of k-NN search, including setting up the field,
+ * query vector, number of neighbors (k), and number of candidates.
+ *
+ * [Elasticsearch reference](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html)
+ */
+class Knn {
+    /**
+     * Creates an instance of Knn.
+     */
+    constructor(field, k, numCandidates) {
+        if (k > numCandidates)
+            throw new Error('Knn numCandidates cannot be less than k');
+        this._body = {};
+        this._body.field = field;
+        this._body.k = k;
+        this._body.filter = [];
+        this._body.num_candidates = numCandidates;
+    }
+
+    /**
+     * Sets the query vector for the k-NN search.
+     * @param {Array<number>} vector - The query vector.
+     * @returns {Knn} Returns the instance of Knn for method chaining.
+     */
+    queryVector(vector) {
+        if (this._body.query_vector_builder)
+            throw new Error(
+                'cannot provide both query_vector_builder and query_vector'
+            );
+        this._body.query_vector = vector;
+        return this;
+    }
+
+    /**
+     * Sets the query vector builder for the k-NN search.
+     * This method configures a query vector builder using a specified model ID and model text.
+     * It's important to note that either a direct query vector or a query vector builder can be
+     * provided, but not both.
+     *
+     * @param {string} modelId - The ID of the model to be used for generating the query vector.
+     * @param {string} modelText - The text input based on which the query vector is generated.
+     * @throws {Error} Throws an error if both query_vector_builder and query_vector are provided.
+     * @returns {Knn} Returns the instance of Knn for method chaining.
+     *
+     * Usage example:
+     * let knn = new Knn();
+     * knn.queryVectorBuilder('model_123', 'Sample model text');
+     */
+    queryVectorBuilder(modelId, modelText) {
+        if (this._body.query_vector)
+            throw new Error(
+                'cannot provide both query_vector_builder and query_vector'
+            );
+        this.query_vector_builder = {
```

Comment:
This should be set on `this._body`, right? The tests are not able to catch this issue so they are not serving the purpose they are meant for. Can you rewrite the tests to not check internal state and instead rely only on the `toJSON` output to validate properties being set?